packages/teleport/src/DesktopSession/DesktopSession.tsx

```diff
@@ -40,42 +46,117 @@ export function DesktopSession(props: State) {
     clipboardState,
     fetchAttempt,
     tdpConnection,
-    wsConnection,
     disconnected,
+    wsConnection,
+    setTdpConnection,
   } = props;
 
-  const clipboardError = clipboardState.enabled && clipboardState.errorText;
-
   const clipboardProcessing =
     clipboardState.enabled && clipboardState.permission.state === 'prompt';
 
-  // Websocket is closed but we haven't
-  // closed it on purpose or registered a tdp error.
-  const unknownConnectionError =
-    wsConnection === 'closed' &&
-    !disconnected &&
-    tdpConnection.status === 'success';
-
   const processing =
     fetchAttempt.status === 'processing' ||
     tdpConnection.status === 'processing' ||
     clipboardProcessing;
 
-  let alertText: string;
-  if (fetchAttempt.status === 'failed') {
-    alertText = fetchAttempt.statusText || 'fetch attempt failed';
-  } else if (tdpConnection.status === 'failed') {
-    alertText = tdpConnection.statusText || 'tdp connection failed';
-  } else if (clipboardError) {
-    alertText = clipboardState.errorText || 'clipboard sharing failed';
-  } else if (unknownConnectionError) {
-    alertText = 'Session disconnected for an unknown reason';
-  }
+  // Manages the state of the error dialog.
+  const [errorDialog, setErrorDialog] = useState({
```

Comment:
This should not be state since this is only a computation result. In current form react will render twice this component
1. render with the initial value
2. render with the computed value (caused by setErrorDialog in `useEffect`)

If you want to group this logic you could use local function. In this case React will render once with the computed value;
```ts
 const computeErrorDialog = () => {
    const clipboardError = clipboardState.enabled && clipboardState.errorText;

    // Websocket is closed but we haven't
    // closed it on purpose or registered a fatal tdp error.
    const unknownConnectionError =
      wsConnection === 'closed' &&
      !disconnected &&
      (tdpConnection.status === 'success' || tdpConnection.status === '');

    let errorText = '';
    if (fetchAttempt.status === 'failed') {
      errorText = fetchAttempt.statusText || 'fetch attempt failed';
    } else if (tdpConnection.status === 'failed') {
      errorText = tdpConnection.statusText || 'tdp connection failed';
    } else if (tdpConnection.status === '') {
      errorText = tdpConnection.statusText || 'encountered a non-fatal error';
    } else if (clipboardError) {
      errorText = clipboardState.errorText || 'clipboard sharing failed';
    } else if (unknownConnectionError) {
      errorText = 'Session disconnected for an unknown reason';
    }
    const open = errorText !== '';
    const fatal = tdpConnection.status !== '';

    return { open, text: errorText, fatal };
}

const errorDialog = computeErrorDialog()
```

Reply:
Good catch